physics/apsides_body.hpp

```diff
@@ -135,6 +137,150 @@ void ComputeApsides(Trajectory<Frame> const& reference,
+                                       &trajectory](Instant const& time) {
+    Position<Frame> const body_position =
+        reference.EvaluatePosition(time);
+    Position<Frame> const trajectory_position =
+        trajectory.EvaluatePosition(time);
+    Displacement<Frame> const displacement =
+        trajectory_position - body_position;
+    return displacement.Norm²();
+  };
+
+  auto const max_radius² = Pow<2>(reference_body.max_radius());
+
+  // Construct the set of all extrema times (apsides and extremities).  In this
+  // set, apoapsides and periapsides alternate.
+  absl::btree_set<Instant> apsides_times{trajectory.t_min(),
+                                         trajectory.t_max()};
+  for (auto const& [time, _] : apoapsides) {
+    apsides_times.insert(time);
+  }
+  for (auto const& [time, _] : periapsides) {
+    apsides_times.insert(time);
+  }
+
+  // Initialize the iterators.  After this block |it| designates the first
+  // periapsis and |previous_it| designates the previous apoapsis, if there is
+  // one, or is past the end.
+  absl::btree_set<Instant>::const_iterator it;
+  absl::btree_set<Instant>::const_iterator previous_it;
+  {
+    auto const first_it = apsides_times.begin();
+    auto const second_it = std::next(first_it);
+    if (squared_distance_from_centre(*first_it) <
+        squared_distance_from_centre(*second_it)) {
+      previous_it = apsides_times.end();
+      it = first_it;
+    } else {
+      previous_it = first_it;
+      it = second_it;
+    }
+  }
+
+  std::vector<Interval<Instant>> intervals;
+  for (; it != apsides_times.end(); previous_it = it, ++it) {
+    // Here |it| designates a periapsis, and |previous_it| the previous
+    // apoapsis, if any.
+    Instant const periapsis_time = *it;
+
+    // No collision is possible if the periapsis is above |max_radius|.
+    if (squared_distance_from_centre(periapsis_time) < max_radius²) {
+      Interval<Instant> interval;
+      if (previous_it == apsides_times.end()) {
+        // No previous periapsis can only happens the first time through the
+        // loop.
+        interval.min = periapsis_time;
+      } else {
+        Instant const apoapsis_time = *previous_it;
+        CHECK_LE(apoapsis_time, periapsis_time);
+
+        if (squared_distance_from_centre(apoapsis_time) > max_radius²) {
+          // The periapsis is below |max_radius| and the preceding apoapsis is
+          // above.  Find the intersection point.
+          interval.min = Brent(
+              [max_radius²,
+               &squared_distance_from_centre](Instant const& time) {
+                return squared_distance_from_centre(time) - max_radius²;
+              },
+              apoapsis_time,
+              periapsis_time);
+        } else {
+          // A periapsis below |max_radius| can only happen the first time
+          // through the loop because the nested loop below skips the other
+          // ones.
+          interval.min = periapsis_time;
```

Comment:
That should be apoapsis, right?
```suggestion
          CHECK_EQ(apoapsis_time, trajectory.t_min());
          interval.min = apoapsis_time;
```

Reply:
Good catch.